dbt/adapters/redshift/connections.py

```diff
@@ -38,33 +40,176 @@ class RedshiftConnectionMethod(StrEnum):
+            # this requirement is really annoying to encode into json schema,
+            # so validate it here
+            if self.credentials.password is None:
+                raise dbt.exceptions.FailedToConnectException(
+                    "'password' field is required for 'database' credentials"
+                )
+
+            def connect():
+                logger.debug("Connecting to redshift with username/password based auth...")
+                c = redshift_connector.connect(
+                    user=self.credentials.user, password=self.credentials.password, **kwargs
+                )
+                if self.credentials.role:
+                    c.cursor().execute("set role {}".format(self.credentials.role))
+                return c
 
+            return connect
 
-class RedshiftConnectionManager(PostgresConnectionManager):
+        elif method == RedshiftConnectionMethod.IAM:
+            if not self.credentials.cluster_id:
+                raise dbt.exceptions.FailedToConnectException(
+                    "Failed to use IAM method, 'cluster_id' must be provided"
+                )
+
+            def connect():
+                logger.debug("Connecting to redshift with IAM based auth...")
+                c = redshift_connector.connect(
+                    iam=True,
+                    db_user=self.credentials.user,
+                    password="",
+                    user="",
+                    cluster_identifier=self.credentials.cluster_id,
+                    profile=self.credentials.iam_profile,
+                    **kwargs,
+                )
+                if self.credentials.role:
+                    c.cursor().execute("set role {}".format(self.credentials.role))
+                return c
+
+            return connect
+        else:
+            raise dbt.exceptions.FailedToConnectException(
+                "Invalid 'method' in profile: '{}'".format(method)
+            )
+
+
+class RedshiftConnectionManager(SQLConnectionManager):
     TYPE = "redshift"
 
+    def _get_backend_pid(self):
+        sql = "select pg_backend_pid()"
+        _, cursor = self.add_query(sql)
+        res = cursor.fetchone()
+        return res
+
+    def cancel(self, connection: Connection):
+        connection_name = connection.name
+        try:
+            pid = self._get_backend_pid()
+            sql = "select pg_terminate_backend({})".format(pid)
+            _, cursor = self.add_query(sql)
+            res = cursor.fetchone()
+            logger.debug("Cancel query '{}': {}".format(connection_name, res))
+        except redshift_connector.error.InterfaceError as e:
+            if "is closed" in str(e):
+                logger.debug(f"Connection {connection_name} was already closed")
+                return
+            raise
+
+    @classmethod
+    def get_response(cls, cursor: redshift_connector.Cursor) -> AdapterResponse:
+        rows = cursor.rowcount
```

Comment:
@colin-rogers-dbt to follow-up on the removal of `cursor.statusmessage` in `get_response`